### PR TITLE
Allow async methods to return Task-like types

### PR DESCRIPTION
**Pull request by cston**
Allow async methods to return types other than `void`, `Task`, or `Task<T>`. Specifically allow `Task`-like types (and `AsyncMethodBuilder` types) if the types contain an expected set of members.

Next: Compare full signature including type constraints when looking up members. (Currently, lookup is by name only.) And add more tests.

[Thanks to Lucian for prototyping this feature.]

### CODE REVIEW

**Comment by cston** [comment]
@dotnet-bot retest this please

**Comment by cston** [comment]
@dotnet-bot retest this please

**Comment by jaredpar** [comment]
CC @roslyn-compiler, @ljw1004

File: src/Compilers/Core/Portable/PublicAPI.Unshipped.txt [hunk 1]
```diff
@@ -660,6 +660,7 @@ Microsoft.CodeAnalysis.Semantics.UnaryOperationKind.UnsignedPrefixIncrement = 77
 abstract Microsoft.CodeAnalysis.Diagnostics.OperationBlockStartAnalysisContext.RegisterOperationAction(System.Action<Microsoft.CodeAnalysis.Diagnostics.OperationAnalysisContext> action, System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.OperationKind> operationKinds) -> void
 abstract Microsoft.CodeAnalysis.Diagnostics.OperationBlockStartAnalysisContext.RegisterOperationBlockEndAction(System.Action<Microsoft.CodeAnalysis.Diagnostics.OperationBlockAnalysisContext> action) -> void
 abstract Microsoft.CodeAnalysis.SemanticModel.GetOperationCore(Microsoft.CodeAnalysis.SyntaxNode node, System.Threading.CancellationToken cancellationToken) -> Microsoft.CodeAnalysis.IOperation
+const Microsoft.CodeAnalysis.WellKnownMemberNames.CreateAsyncMethodBuilder = "CreateAsyncMethodBuilder" -> string
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/PublicAPI.Unshipped.txt]: Do we mean for this to be public?

File: src/Compilers/Core/Portable/Symbols/WellKnownMemberNames.cs [hunk 2]
```diff
@@ -298,5 +298,11 @@ public static class WellKnownMemberNames
         /// (see C# Specification, §7.7.7.1 Awaitable expressions).
         /// </summary>
         public const string OnCompleted = nameof(OnCompleted);
+
+        /// <summary>
+        /// The required name for the <c>CreateAsyncMethodBuilder</c> method used to
+        /// build a tasklike instance.
+        /// </summary>
+        public const string CreateAsyncMethodBuilder = "CreateAsyncMethodBuilder";
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/Symbols/WellKnownMemberNames.cs]: > public const string CreateAsyncMethodBuilder [](start = 7, length = 45)

WRT my earlier comment about being public: looks like this is the sttandard pattern we follow here.

File: src/Compilers/CSharp/Portable/Binder/Binder_Await.cs [hunk 3]
```diff
@@ -431,5 +432,102 @@ private static bool HasOptionalOrVariableParameters(MethodSymbol method)
 
             return false;
         }
+
+        /// <summary>
+        /// Returns the appropriate AsyncMethodBuilder based on whether the
+        /// return type is void or a non-generic or generic task-like type. The
+        /// caller is responsible for ensuring this is used for async methods only.
+        /// </summary>
+        internal AsyncMethodBuilder GetAsyncMethodBuilder(CSharpSyntaxNode syntax, TypeSymbol returnType)
```
Comment (ljw1004) [inline src/Compilers/CSharp/Portable/Binder/Binder_Await.cs]: Could you explain conceptually why GetAsyncMethodBuilder is inside "binder", specifically the part of binder that's responsible for turning Await into a BoundAwaitExpression?

Because the binding for await expressions is completely unrelated to async-method-builder, I think...?

File: src/Compilers/CSharp/Portable/Symbols/AsyncKind.cs [hunk 4]
```diff
@@ -0,0 +1,13 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols
+{
+    internal enum AsyncKind : byte
+    {
+        NotAsync = 0,
+        Invalid,
+        Void,
+        Task,
```
Comment (ljw1004) [inline src/Compilers/CSharp/Portable/Symbols/AsyncKind.cs]: I'd rename these... the hardest thing in this feature was recognizing whether we're talking about "Task" specifically or "non-generic Tasklike", and likewise about "Task<T>" specifically or "generic Tasklike".

In this case, the code uses AsyncKind.Task for all non-generic tasklikes, and uses ASyncKind.Task_T for all generic tasklikes. So I'd hope for a name that indicates this :)

File: src/Compilers/CSharp/Portable/Binder/Binder_Await.cs (hunk 3 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Await.cs]: The method is defined on `Binder` to allow using `Binder.LookupMembersInType` to find the `CreateAsyncMethodBuilder` method, in particular to use the accessibility checking of `LookupMembersInType`. It should probably not be grouped with the partial class for Awaiter though - I'll move it.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 5]
```diff
@@ -6799,7 +6799,7 @@ private bool ContainingMethodOrLambdaRequiresValue()
             return
                 (object)containingMethod == null ||
                     !containingMethod.ReturnsVoid &&
-                    !containingMethod.IsTaskReturningAsync(this.Compilation);
+                    containingMethod.AsyncKind != AsyncKind.Task;
         }
```
Comment (ljw1004) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: The logic of this method is

* non-async, void -> false
* non-async, non-void -> true
* async, AsyncKind.NotAsync -> (I trust this is impossible??)
* async, Invalid -> true
* async, Void -> (already caught by containingMethod.ReturnsVoid)
* async, Task -> false
* async, Task_T -> true

I wonder about that AsyncKind.Invalid case. Should it maybe return FALSE from this method `ContainingMethodRequiresValue` ? (I can't think of an example...)

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 6]
```diff
@@ -3115,14 +3114,22 @@ protected bool IsInAsyncMethod()
 
         protected bool IsTaskReturningAsyncMethod()
```
Comment (ljw1004) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: I preferred my naming convention because it was less subtle :) even at the cost of wordiness
`IsNongenericTaskOrTasklikeReturningAsyncMethod`
`IsGenericTaskOrTasklikeReturningAsyncMethod`

File: src/Compilers/CSharp/Portable/Symbols/AsyncKind.cs (hunk 4 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/AsyncKind.cs]: I actually liked the short form `Task` to mean `Task` or `Task`-like (similarly for `Task_T`), particularly since the consumers of this enum are typically not interested in the distinction. Any suggestions on names? `TaskOrTasklike` seems clumsy.

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs [hunk 7]
```diff
@@ -132,10 +146,17 @@ private static TypeSymbol InferReturnType(BoundBlock block, Binder binder, bool
...
             // Async:
 
+            var delegateReturnType = delegateType?.GetDelegateType()?.DelegateInvokeMethod?.ReturnType as NamedTypeSymbol;
+            var asyncMethodBuilder = (object)delegateReturnType == null ?
+                AsyncMethodBuilder.NotAsync :
+                binder.GetAsyncMethodBuilder(block.Syntax, delegateReturnType);
+
             if (resultTypes.IsEmpty)
             {
                 // No return statements have expressions; inferred type Task:
```
Comment (ljw1004) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs]: Adjust this comment "inferred type either Task or, if delegate was supplied, ..."

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs [hunk 8]
```diff
@@ -428,22 +464,16 @@ private BoundLambda ReallyBind(NamedTypeSymbol delegateType)
...
             {
-                if ((object)returnType != null && // Can be null if "delegateType" is not actually a delegate type.
-                    returnType.SpecialType != SpecialType.System_Void &&
-                    returnType != binder.Compilation.GetWellKnownType(WellKnownType.System_Threading_Tasks_Task) &&
-                    returnType.OriginalDefinition != binder.Compilation.GetWellKnownType(WellKnownType.System_Threading_Tasks_Task_T))
-                {
-                    // Cannot convert async {0} to delegate type '{1}'. An async {0} may return void, Task or Task&lt;T&gt;, none of which are convertible to '{1}'.
-                    diagnostics.Add(ErrorCode.ERR_CantConvAsyncAnonFuncReturns, lambdaSymbol.Locations[0], lambdaSymbol.MessageID.Localize(), delegateType);
-                }
+                // Cannot convert async {0} to delegate type '{1}'. An async {0} may return void, Task or Task&lt;T&gt;, none of which are convertible to '{1}'.
```
Comment (ljw1004) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs]: We have to update this error message text, yes?

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs (hunk 7 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs]: Modified comment here and below.

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs (hunk 8 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs]: We have several diagnostic messages that mention `Task` or `Task<T>`: 
```
ERR_BadAsyncReturnExpression
ERR_CantConvAsyncAnonFuncReturns
ERR_TaskRetNoObjectRequired
ERR_BadAsyncReturn
WRN_UnobservedAwaitableExpression_Description
```

File: src/Compilers/CSharp/Portable/Binder/Binder_Await.cs (hunk 3 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Await.cs]: Changed to no longer use `Binder.LookupMembersInType`. Now lookup simply uses `NamedTypeSymbol.GetMembers(name)`.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 5 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: This change has been reverted.

**Comment by cston** [comment]
@ljw1004, @dotnet/roslyn-compiler The PR has been updated with the changes to overload resolution and tests ported from Lucian's `features/async-return` branch. There are still a number of `// PROTOTYPE()` comments that I am working on now.

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs [hunk 9]
```diff
@@ -1252,9 +1252,12 @@ private static TypeSymbol GetParameterType(int argIndex, MemberAnalysisResult re
                                                        out okToDowngradeToNeither);
                 }
 
+                var type1Normalized = type1.NormalizeTaskTypes(Compilation);
+                var type2Normalized = type2.NormalizeTaskTypes(Compilation);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs]: Note `NormalizeTaskTypes()` is called in `OverloadResolution.BetterFunctionMember<>` whenever comparing parameter types to ignore differences in `Task` types. And `NormalizeTaskTypes()` walks the type calling `GetAsyncMethodBuilderType()` on each type part, and substituting `System.Threading.Tasks.Task` or `System.Threading.Tasks.Tasks<T>` for any `Task`-like types. However, the overhead should by low since `GetAsyncMethodBuilderType()` is essentially a dictionary lookup, and substitution only occurs if `Task`-like types are found.

**Comment by cston** [comment]
@ljw1004, @dotnet/roslyn-compiler, ping, please review.

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs (hunk 9 above)
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs]: What happens in the case where `Task` and `Task<T>` are not available?  Does it break our ability to do overload resolution on other task-like types?

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs [hunk 10]
```diff
@@ -2096,28 +2092,26 @@ public override BoundNode VisitReturnStatement(BoundReturnStatement node)
...
+                                                      ((NamedTypeSymbol)type2).TypeArgumentsNoUseSiteDiagnostics[0],
+                                                      ref useSiteDiagnostics, betterConversionTargetRecursionLimit);
                 }
+
+                // A shortcut, Task<T> type cannot satisfy other rules.
+                return BetterResult.Neither;
+            }
+            else if (type2IsGenericTask)
+            {
+                // A shortcut, Task<T> type cannot satisfy other rules.
```
Comment (ljw1004) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs]: You can only keep this shortcut if a stronger invariant holds, that NO TASKLIKE<T> can satisfy the other rules. I'm not sure what are the "other rules" mentioned by this comment. Could you check?

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs (hunk 9 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs]: Good catch. If `Task` and `Task<T>` are not available, we are currently substituting an error type. I've changed `NormalizeTaskTypes` to skip substitution in that case but that means we will not consider two different task-like types as equivalent if the well known types are missing.

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs [hunk 11]
```diff
@@ -132,10 +146,25 @@ private static TypeSymbol InferReturnType(BoundBlock block, Binder binder, bool
 
             // Async:
 
+            NamedTypeSymbol taskType = null;
```
Comment (ljw1004) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs]: Could you comment what this chunk of code is doing? It's not obvious. I think it says:
// This lambda has an "async" modifier so we need to figure out what tasklike the lambda returns.
// The default assumption is it will return Task or Task<T> (depending on whether it has returns or not).
// But if the lambda is being matched against a delegate type (i.e. not a generic type parameter, and
// not a non-delegate type), and if the delegate type has a return type, and if that return type is
// tasklike (as identified by the presence of a createBuilderMethod), then the delegate's return type
// indicates which tasklike the lambda should return. It might be Task, or Task<T>, or Task<concrete>,
// or Tasklike, or Tasklike<T>, or Tasklike<concrete>.

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs [hunk 12]
```diff
@@ -132,10 +146,25 @@ private static TypeSymbol InferReturnType(BoundBlock block, Binder binder, bool
...
+            var delegateReturnType = delegateType?.GetDelegateType()?.DelegateInvokeMethod?.ReturnType as NamedTypeSymbol;
+            if ((object)delegateReturnType != null)
+            {
+                MethodSymbol createBuilderMethod;
+                var delegateTaskBuilderType = delegateReturnType.GetAsyncMethodBuilderType(out createBuilderMethod);
```
Comment (ljw1004) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs]: Might delegateReturnType be null in case of a void-returning delegate, or error, or other reason?

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs [hunk 13]
```diff
@@ -132,10 +146,25 @@ private static TypeSymbol InferReturnType(BoundBlock block, Binder binder, bool
 
             // Async:
 
+            NamedTypeSymbol taskType = null;
+            var delegateReturnType = delegateType?.GetDelegateType()?.DelegateInvokeMethod?.ReturnType as NamedTypeSymbol;
```
Comment (ljw1004) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs]: There's also the method "DelegateReturnType" you could have used

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs [hunk 14]
```diff
@@ -341,33 +374,35 @@ internal IEnumerable<TypeSymbol> InferredReturnTypes()
             }
         }
 
-        private static ImmutableArray<ParameterSymbol> DelegateParameters(NamedTypeSymbol delegateType)
```
Comment (ljw1004) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs]: Just a personal preference, but I don't really like this factoring out into doing everything via the delegateInvokeMethod. It feels more clear to me if all functions operate upon delegateType. (it feels kind of underhand to use the invokeMethod as a proxy/cached-lookup of the important information on the delegateType).

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs [hunk 15]
```diff
@@ -341,33 +374,35 @@ internal IEnumerable<TypeSymbol> InferredReturnTypes()
...
         {
-            NamedTypeSymbol d = delegateType.GetDelegateType();
-            if ((object)d == null || (object)d.DelegateInvokeMethod == null || d.DelegateInvokeMethod.ReturnsVoid)
+            if ((object)invokeMethod == null || invokeMethod.ReturnsVoid)
             {
                 return false;
             }
 
-            if (IsAsync && this.binder.Compilation.GetWellKnownType(WellKnownType.System_Threading_Tasks_Task) == d.DelegateInvokeMethod.ReturnType)
+            if (IsAsync && invokeMethod.ReturnType.IsNonGenericTaskType(this.binder.Compilation))
```
Comment (ljw1004) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs]: Again, is ReturnType always non-null?

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs [hunk 16]
```diff
@@ -510,22 +567,18 @@ public BoundLambda BindForReturnTypeInference(NamedTypeSymbol delegateType)
...
-                }
+                return invoke;
             }
 
             // delegateType or DelegateInvokeMethod can be null in cases of malformed delegates
-            // in such case we would want something trivial with no parameters, like a fake static ctor
-            return new SynthesizedStaticConstructor(delegateType);
+            // in such case we would want something trivial with no parameters, like a fake static ctor.
+            // Since the containingType of the .cctor must be non-null, System.Object is used.
+            return new SynthesizedStaticConstructor(binder.Compilation.GetSpecialType(SpecialType.System_Object));
```
Comment (ljw1004) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs]: I can't find just now the place we're you're generating the cache. I had changed the cache to also key upon return type for ALL lambdas, even though keying upon return type is only needed for async lambdas. I worried that this would hurt perf in the common non-async case. Did you do the same? have the same worry?

File: src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodBuilderMemberCollection.cs [hunk 17]
```diff
@@ -95,84 +103,69 @@ internal static bool TryCreate(SyntheticBoundNodeFactory F, MethodSymbol method,
...
+                    createBuilderMethod: null,
+                    createBuilder: WellKnownMember.System_Runtime_CompilerServices_AsyncVoidMethodBuilder__Create,
                     setException: WellKnownMember.System_Runtime_CompilerServices_AsyncVoidMethodBuilder__SetException,
                     setResult: WellKnownMember.System_Runtime_CompilerServices_AsyncVoidMethodBuilder__SetResult,
                     awaitOnCompleted: WellKnownMember.System_Runtime_CompilerServices_AsyncVoidMethodBuilder__AwaitOnCompleted,
                     awaitUnsafeOnCompleted: WellKnownMember.System_Runtime_CompilerServices_AsyncVoidMethodBuilder__AwaitUnsafeOnCompleted,
                     start: WellKnownMember.System_Runtime_CompilerServices_AsyncVoidMethodBuilder__Start_T,
                     setStateMachine: WellKnownMember.System_Runtime_CompilerServices_AsyncVoidMethodBuilder__SetStateMachine,
-                    task: null,
+                    task: WellKnownMember.Count,
```
Comment (ljw1004) [inline src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodBuilderMemberCollection.cs]: Deserves a comment if you're using this as a sentinel! I wonder if you might maybe have a constant with this value but a more descriptive name?

File: src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodBuilderMemberCollection.cs [hunk 18]
```diff
@@ -183,39 +176,47 @@ internal static bool TryCreate(SyntheticBoundNodeFactory F, MethodSymbol method,
...
+            MethodSymbol createBuilderMethod,
+            WellKnownMember createBuilder,
             WellKnownMember setException,
             WellKnownMember setResult,
             WellKnownMember awaitOnCompleted,
             WellKnownMember awaitUnsafeOnCompleted,
             WellKnownMember start,
             WellKnownMember setStateMachine,
-            PropertySymbol task,
+            WellKnownMember task,
```
Comment (ljw1004) [inline src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodBuilderMemberCollection.cs]: Oh, I see why you can no longer use null.

File: src/Compilers/CSharp/Portable/Symbols/MemberSignatureComparer.cs [hunk 19]
```diff
@@ -239,7 +239,7 @@ internal class MemberSignatureComparer : IEqualityComparer<Symbol>
         public static readonly MemberSignatureComparer LambdaReturnInferenceCacheComparer = new MemberSignatureComparer(
             considerName: false,                // valid invoke is always called "Invoke"
             considerExplicitlyImplementedInterfaces: false,
-            considerReturnType: false,          // do not care
+            considerReturnType: true,          // to differentiate Task types
```
Comment (ljw1004) [inline src/Compilers/CSharp/Portable/Symbols/MemberSignatureComparer.cs]: Ah, here's the caching key thing I was concerned about perf earlier.

File: src/Compilers/CSharp/Portable/Symbols/Source/LambdaSymbol.cs [hunk 20]
```diff
@@ -1,21 +1,25 @@
...
 using System.Collections.Generic;
 using System.Collections.Immutable;
-using System.Diagnostics;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.CSharp.Symbols
 {
     internal sealed class LambdaSymbol : MethodSymbol
     {
+        internal delegate void GetReturnTypeDelegate(
```
Comment (ljw1004) [inline src/Compilers/CSharp/Portable/Symbols/Source/LambdaSymbol.cs]: Oh! So this now has a delegate allocation cost each time. I wonder why you switched from computing these things up-front to computing them lazily?

File: src/Compilers/CSharp/Portable/Symbols/Source/LambdaSymbol.cs [hunk 21]
```diff
@@ -196,17 +183,6 @@ public override TypeSymbol ReturnType
             get { return _returnType; }
         }
 
-        // In error recovery and type inference scenarios we do not know the return type
```
Comment (ljw1004) [inline src/Compilers/CSharp/Portable/Symbols/Source/LambdaSymbol.cs]: Is this related to why you switched to doing it lazily?

File: src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs [hunk 22]
```diff
@@ -2,34 +2,65 @@
...
-        private ImmutableArray<TypeParameterConstraintClause> _lazyTypeParameterConstraints;
         private readonly RefKind _refKind;
-        private TypeSymbol _returnType;
-        private bool _isVararg;
+        private ParametersAndDiagnostics _lazyParametersAndDiagnostics;
+        private TypeParameterConstraintsAndDiagnostics _lazyTypeParameterConstraintsAndDiagnostics;
+        private ReturnTypeAndDiagnostics _lazyReturnTypeAndDiagnostics;
         private TypeSymbol _iteratorElementType;
-
-        // TODO: Find a better way to report diagnostics.
```
Comment (ljw1004) [inline src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs]: Is this "better way" change related to Tasklike?

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs [hunk 23]
```diff
@@ -1240,5 +1240,154 @@ private static bool HasNamespaceName(NamespaceSymbol @namespace, string namespac
...
+            return (object)builderType != null;
+        }
+
+        /// <summary>
+        /// Returns the async method builder type for generic and non-generic task-like types.
+        /// </summary>
+        internal static NamedTypeSymbol GetAsyncMethodBuilderType(this NamedTypeSymbol type, out MethodSymbol createBuilderMethod)
+        {
+            Debug.Assert((object)type != null);
+            Debug.Assert(type.SpecialType != SpecialType.System_Void);
```
Comment (ljw1004) [inline src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs]: I'm surprised at this assert. I think you might legitimately call GetASyncMehtodBUilderType upon void, no?

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs [hunk 24]
```diff
@@ -1240,5 +1240,154 @@ private static bool HasNamespaceName(NamespaceSymbol @namespace, string namespac
...
+            var constructedFrom = type.ConstructedFrom;
+            if ((object)constructedFrom == type)
+            {
+                MethodSymbol createBuilderMethod;
+                var builderType = type.GetAsyncMethodBuilderType(out createBuilderMethod);
+                if ((object)builderType == null)
+                {
+                    return false;
+                }
+                Debug.Assert(type.Arity < 2);
```
Comment (ljw1004) [inline src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs]: Why this assert? It looks like GetASyncMethodBuilderType makes no test of the arity of the tasklike.

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs [hunk 25]
```diff
@@ -1240,5 +1240,154 @@ private static bool HasNamespaceName(NamespaceSymbol @namespace, string namespac
...
+                    type.Arity == 1 ?
+                    WellKnownType.System_Threading_Tasks_Task_T :
+                    WellKnownType.System_Threading_Tasks_Task);
+                return true;
+            }
+            else
+            {
+                var typeArgs = type.TypeArguments;
+                var builder = ArrayBuilder<TypeSymbol>.GetInstance();
+                bool hasChanged = false;
```
Comment (ljw1004) [inline src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs]: I wonder if it'd perform better to run through just doing a check, and if it returns false then skip the builder.

Only allocate the builder in the (rare) case where it is tasklike. This will be slower in the non-Task tasklike case, but involve less memory allocation in the much more common non-tasklike case.

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs (hunk 9 above)
Comment (ljw1004) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs]: That's funny! Even if we fix this particular problem (to consider two different task-like types as equivalent even in the absence of System.Threading.Tasks.Task/Task<T>), the compiler is still hard-coded to prefer Task in certain situations e.g.

```csharp
void f<T>(Func<T> lambda)
f(async () => 3);
```

I think it's a mistake that the compiler does this inference, but we're stuck with it. Imagine if we came up with a set of libraries which removed Task completely, and only used other tasks, as a way to force the compiler never to do these dangerous inferences!!!

Still it's a far-removed way-out scenario. Task has found its way into the deepest part of the BCL and will never be removed.

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs (hunk 13 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs]: Unfortunately `DelegateReturnType` is defined on `UnboundLambdaState` not `BoundLambda`.

File: src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodBuilderMemberCollection.cs (hunk 17 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodBuilderMemberCollection.cs]: Added a comment.

File: src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs (hunk 22 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs]: The change was needed in an earlier iteration of this PR but is no longer strictly necessary.

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs (hunk 11 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs]: Added comment.

File: src/Compilers/CSharp/Portable/Symbols/Source/LambdaSymbol.cs (hunk 20 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/LambdaSymbol.cs]: This was left over from an earlier iteration. Reverted.

File: src/Compilers/CSharp/Portable/Symbols/Source/LambdaSymbol.cs (hunk 21 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/LambdaSymbol.cs]: Reverted.

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs (hunk 23 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs]: Changed the name and signature of the method to your suggestion and checked arity.

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs (hunk 24 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs]: Added check to `GetAsyncMethodBuilderType`, now `IsCustomTaskType`.

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs (hunk 10 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs]: The remaining rules are for delegate types and integral types.

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs (hunk 12 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs]: Thanks, I will add tests for void-returning and error delegate types in a separate PR.

**Comment by cston** [comment]
Updated PR checked in to `dotnet:dev15-preview-4`. See https://github.com/dotnet/roslyn/pull/12518.

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs (hunk 8 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs]: Logged #12621.
